File: src/xmss-alt/hash_address.c
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
// Based on the public domain XMSS reference implementation
// by Andreas Hülsing and Joost Rijneveld
/*
hash_address.c version 20160722
Andreas Hülsing
Joost Rijneveld
Public domain.
*/
#include <cstdint>

// TODO: Replace all this with a union

void setLayerADRS(uint32_t adrs[8], uint32_t layer) {
    adrs[0] = layer;
}

void setTreeADRS(uint32_t adrs[8], uint64_t tree) {
    adrs[1] = (uint32_t) (tree >> 32);
    adrs[2] = (uint32_t) tree;
}

void setType(uint32_t adrs[8], uint32_t type) {
    adrs[3] = type;
    int i;
    for (i = 4; i < 8; i++) {
        adrs[i] = 0;
    }
}

void setKeyAndMask(uint32_t adrs[8], uint32_t keyAndMask) {
    adrs[7] = keyAndMask;
}

// OTS

void setOTSADRS(uint32_t adrs[8], uint32_t ots) {
    adrs[4] = ots;
}

void setChainADRS(uint32_t adrs[8], uint32_t chain) {
    adrs[5] = chain;
}

void setHashADRS(uint32_t adrs[8], uint32_t hash) {
    adrs[6] = hash;
}

// L-tree

void setLtreeADRS(uint32_t adrs[8], uint32_t ltree) {
    adrs[4] = ltree;
}

// Hash Tree & L-tree

void setTreeHeight(uint32_t adrs[8], uint32_t treeHeight) {
    adrs[5] = treeHeight;
}

void setTreeIndex(uint32_t adrs[8], uint32_t treeIndex) {
    adrs[6] = treeIndex;
}

File: src/rust_wrapper/xmss_alt/hash_address.rs
pub fn set_layer_adrs(adrs: &mut [u32], layer: u32) {
    adrs[0] = layer;
}

pub fn set_tree_adrs(adrs: &mut [u32], tree: u64) {
    adrs[1] = (tree >> 32) as u32;
    adrs[2] = tree as u32;
}

pub fn set_type(adrs: &mut [u32], type_t: u32) {
    adrs[3] = type_t;
    for i in 4..8 {
        adrs[i] = 0;
    }
}

pub fn set_key_and_mask(adrs: &mut [u32], key_and_mask: u32) {
    adrs[7] = key_and_mask;
}

// OTS

pub fn set_ots_adrs(adrs: &mut [u32], ots: u32) {
    adrs[4] = ots;
}

pub fn set_chain_adrs(adrs: &mut [u32], chain: u32) {
    adrs[5] = chain;
}

pub fn set_hash_adrs(adrs: &mut [u32], hash: u32) {
    adrs[6] = hash;
}

// L-tree

pub fn set_ltree_adrs(adrs: &mut [u32], ltree: u32) {
    adrs[4] = ltree;
}

// Hash Tree & L-tree

pub fn set_tree_height(adrs: &mut [u32], tree_height: u32) {
    adrs[5] = tree_height;
}

pub fn set_tree_index(adrs: &mut [u32], tree_index: u32) {
    adrs[6] = tree_index;
}
